docs/reconfigurator.adoc

```diff
@@ -0,0 +1,183 @@
+
+NOTE: Currently, these "user"-driven steps are carried out using `omdb`, which means that in production systems they can really only be done by an Oxide employee with access via the technician port.  The long-term goal is to automate all of this so that planning and execution both happen continuously and automatically.
+
+You might notice that the only step that has anything to do with adding a sled is the first one.  This is where a user communicated a change to the **intended state** of the world.  Everything Reconfigurator does follows a pattern just like this:
+
+1. A user changes the intended state of the world.
+2. The user invokes the planner to generate a new blueprint.
+3. The user makes the new blueprint the current target.
+4. Reconfigurator executes the target blueprint.
+
+Reconfigurator makes heavy use of well-established patterns, most notably:
+
+* The "Reconciler" pattern (which we've elsewhere called the "Reliable Persistent Workflow" or "RPW" pattern -- see <<rfd373>>).  In this pattern, a system is constantly comparing an intended state of the world against the actual state, determining what to do to bring them in line, and then doing it.  **This is very different from distributed sagas, where the set of steps to be taken is fixed up front and cannot change if either the intended state or the actual state changes.**  Reconcilers are not necessarily _better_ than distributed sagas.  Each is suited to different use cases.
+* The <<plan-execute-pattern>>, in which complex operations are separated into a planning phase and an execution phase.  This separation is fundamental to Reconfigurator.
+
+== Reconfigurator overview
+
+Important data:
+
+* **Policy** describes _intended_ system state in terms that a user or Oxide technician might want to control: what release the system is running, what sleds are part of the system, how many CockroachDB nodes there should be, etc.
+* **Inventory collections** describe the actual state of the system, particularly what hardware exists, what versions of what software components are running everywhere, the software's configuration, etc.
+* **Blueprints** are very detailed descriptions of the intended state of the system.  Blueprints are stored in CockroachDB.  They are described in detail below.
+* At any given time, there is one **target blueprint** that the system is continuously trying to make real.  This is stored in CockroachDB.
+
+Reconfigurator consists mainly of:
+
+* a **planner** that takes as input the latest policy and the latest state of the system (including inventory collections and other database state) and produces a **blueprint**.  Generating a blueprint does not make it the new target!  The new blueprint is a _possible_ next state of the world.  Only when it's made the current target does the system attempt to make that blueprint real.
```

Comment:
Since there's been some confusion on this point, it's maybe worth expanding this:

* A newly-generated blueprint may _never_ become the target
* Therefore, the planner cannot change the state of the world (e.g., when the planner makes allocation decisions like picking IP addresses, it can't run database queries to actually allocate those IPs). Instead, it makes optimistic decisions and expects the executor to actually acquire resources, update zones, etc.

Reply:
Updated in 4bed612b900bb556e8cd3ec97d2394ea8d8f0ec5.